packages/Search/src/details/DTK_DetailsDistributedSearchTreeImpl.hpp

```diff
@@ -0,0 +1,535 @@
+
+    Kokkos::Experimental::MinMaxScalar<int> result;
+    Kokkos::Experimental::MinMax<int> reducer( result );
+    parallel_reduce(
+        Kokkos::RangePolicy<ExecutionSpace>( 0, n ),
+        Kokkos::Impl::min_max_functor<Kokkos::View<int *, DeviceType>>(
+            query_ids ),
+        reducer );
+    if ( result.min_val == result.max_val )
+        return;
+    Kokkos::BinSort<Kokkos::View<int *, DeviceType>, CompType> bin_sort(
+        query_ids, CompType( n / 2, result.min_val, result.max_val ), true );
+    bin_sort.create_permute_vector();
+    bin_sort.sort( results );
+    bin_sort.sort( ranks );
+    if ( distances_ptr )
+        bin_sort.sort( *distances_ptr );
+    Kokkos::fence();
+}
+
+template <typename DeviceType>
+void DistributedSearchTreeImpl<DeviceType>::count_results(
+    int n_queries, Kokkos::View<int *, DeviceType> query_ids,
+    Kokkos::View<int *, DeviceType> &offset )
+{
+    int const nnz = query_ids.extent( 0 );
+
+    Kokkos::realloc( offset, n_queries + 1 );
+    fill( offset, 0 );
+
+    Kokkos::parallel_for(
+        REGION_NAME( "count_results_per_query" ),
+        Kokkos::RangePolicy<ExecutionSpace>( 0, nnz ), KOKKOS_LAMBDA( int i ) {
+            Kokkos::atomic_increment( &offset( query_ids( i ) ) );
+        } );
+    Kokkos::fence();
+
+    exclusive_prefix_sum( offset );
+}
+
+template <typename DeviceType>
+template <typename Query>
+void DistributedSearchTreeImpl<DeviceType>::forward_queries(
+    Teuchos::RCP<Teuchos::Comm<int> const> comm,
+    Kokkos::View<Query *, DeviceType> queries,
+    Kokkos::View<int *, DeviceType> indices,
+    Kokkos::View<int *, DeviceType> offset,
+    Kokkos::View<Query *, DeviceType> &fwd_queries,
+    Kokkos::View<int *, DeviceType> &fwd_ids,
+    Kokkos::View<int *, DeviceType> &fwd_ranks )
+{
+    int const comm_rank = comm->getRank();
+
+    Tpetra::Distributor distributor( comm );
+
+    int const n_queries = queries.extent( 0 );
+    int const n_exports = offset( n_queries );
+    int const n_imports = distributor.createFromSends(
+        Teuchos::ArrayView<int>( indices.data(), n_exports ) );
+
+    Kokkos::View<Query *, DeviceType> exports( queries.label(), n_exports );
+    Kokkos::parallel_for( REGION_NAME( "forward_queries_fill_buffer" ),
+                          Kokkos::RangePolicy<ExecutionSpace>( 0, n_queries ),
+                          KOKKOS_LAMBDA( int q ) {
+                              for ( int i = offset( q ); i < offset( q + 1 );
+                                    ++i )
+                              {
+                                  exports( i ) = queries( q );
+                              }
+                          } );
+    Kokkos::fence();
+
+    Kokkos::View<int *, DeviceType> export_ranks( "export_ranks", n_exports );
+    fill( export_ranks, comm_rank );
+
+    Kokkos::View<int *, DeviceType> import_ranks( "import_ranks", n_imports );
+    send_across_network( distributor, export_ranks, import_ranks );
```

Comment:
This should give you the amount of stuff you are sending/receiving:

```
    /// \brief Number of values this process will receive from each process.
    ///
    /// This process will receive <tt>getLengthsFrom[i]</tt> values
    /// from process <tt>getProcsFrom[i]</tt>.
    ///
    /// This is a nonpersisting view.  It will last only as long as
    /// this Distributor instance does.
    Teuchos::ArrayView<const size_t> getLengthsFrom() const;

    /// \brief Number of values this process will send to each process.
    ///
    /// This process will send <tt>getLengthsTo[i]</tt> values to
    /// process <tt>getProcsTo[i]</tt>.
    ///
    /// This is a nonpersisting view.  It will last only as long as
    /// this Distributor instance does.
    Teuchos::ArrayView<const size_t> getLengthsTo() const;
```

So based on this I would say I was incorrectly sending ranks in the old code as well as you can reconstruct that information with on-process data